Renaming Filters

alexkim205 opened the pull request:
## Changes

Adds the ability to rename entity filters for all of our insights. This PR implements the frontend side of this. All new features are hidden behind a FF called `6063-rename-filters`.

Backend PR: #6090

Closes #5229.
Closes #5207.

~~Note: This PR will be rebased on `master` once #6090 is in. You can ignore diffs in all *.py files~~

## Screenshots

**Demo**

https://user-images.githubusercontent.com/13460330/134576038-ead070e8-2a56-48ca-88a5-bdcc0521ea04.mov


A few edge cases below. Note, I probably haven't tested all edge cases so make sure to click around and try to break things.

**Trends + Legend**

<img width="1294" alt="Screen Shot 2021-09-23 at 12 41 09 PM" src="https://user-images.githubusercontent.com/13460330/134574805-4aac70bc-4410-4921-9efa-2da36bdbb57b.png">

**Trends table**

<img width="1298" alt="Screen Shot 2021-09-23 at 12 41 13 PM" src="https://user-images.githubusercontent.com/13460330/134574809-b65f90d9-d224-4634-bcab-0d29348c3f58.png">

**Dashboards**

<img width="1009" alt="Screen Shot 2021-09-23 at 12 57 43 PM" src="https://user-images.githubusercontent.com/13460330/134575263-fd83f940-3fb6-473b-b2f7-87995df2c15d.png">

**Saved Insight**

<img width="1329" alt="Screen Shot 2021-09-23 at 12 57 13 PM" src="https://user-images.githubusercontent.com/13460330/134575224-72a2a579-bcd3-453e-a3ad-074d8f44cc6e.png">


**With vertical breakdowns**

<img width="1136" alt="Screen Shot 2021-09-23 at 12 40 54 PM" src="https://user-images.githubusercontent.com/13460330/134574801-e85afb81-d1b6-43f7-8bde-5b2c3aa31f0c.png">

**Without vertical breakdowns**

<img width="1304" alt="Screen Shot 2021-09-23 at 12 44 02 PM" src="https://user-images.githubusercontent.com/13460330/134574816-e4c912e6-a2ba-46d4-b5af-018dad202ba0.png">



## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [x] Jest frontend tests
- [x] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [x] New/changed UI is decent on smartphones (viewport width around 360px)

alexkim205 committed:
add custom_name to model

clarkus reviewed (commented):
I tried using the deployment to review these changes, but doesn't quite seem to be working. The modal never triggers to facilitate the renaming.

alexkim205 commented:
> I tried using the deployment to review these changes, but doesn't quite seem to be working. The modal never triggers to facilitate the renaming.

Ah sorry Chris, this is still in draft mode so I haven't implemented the modal yet. I should've mentioned this in the CR description. I'll bump you again once it's ready for review, sorry for the inconvenience!

clarkus commented:
No worries. Thanks!

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 committed:
added testing and functionality

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 committed:
add custom_name to entity model

alexkim205 committed:
Merge branch 'feat/entity-filter-renaming-backend' into feat/entity-filter-renaming

alexkim205 committed:
get new entity info working on legends, tables, and other contexts

alexkim205 committed:
fix test and add custom name to funnel

alexkim205 committed:
fix more tests

alexkim205 committed:
Merge branch 'feat/entity-filter-renaming-backend' into feat/entity-filter-renaming

alexkim205 committed:
finish up funnel edge cases

alexkim205 committed:
remaining broken tests

alexkim205 committed:
Merge branch 'feat/entity-filter-renaming-backend' into feat/entity-filter-renaming

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 commented:
Ready! retagging for reviews @clarkus @mariusandra @Twixes

alexkim205 committed:
update action in logic test

clarkus reviewed (dismissed):
Looks good to me! The one change I think we should consider before merging is shown in this screenshot. When you trigger the rename, you're presented with the modal and the name field. There's a default value there. I think if we focused and selected the field value, it'd be easy for a user to use key commands to replace, amend, whatever the text there. It would just be one less action needed in order to rename the step. 

<img width="300" alt="Screen Shot 2021-09-23 at 3 17 03 PM" src="https://user-images.githubusercontent.com/254612/134591356-c0f3234d-064e-40e5-b7c4-c6706a0f4f44.png">

alexkim205 committed:
remove consoles

alexkim205 committed:
autofocus

alexkim205 commented:
Autofocusing now works

https://user-images.githubusercontent.com/13460330/134596033-14ee2756-316e-46cf-a11b-583767500df1.mov

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

rcmarron commented on frontend/src/scenes/insights/utils.ts:
```diff
@@ -19,3 +19,12 @@ export const getLogicFromInsight = (
+    filter: EntityFilter | ActionFilter | null,
+    isCustom = true
+): string | null => {
+    const customName = (filter?.custom_name ?? '').trim() === '' ? null : filter?.custom_name
```
Nit: this took me a bit to understand what was happening. Might be good to break it out into a few lines.

rcmarron reviewed (dismissed):
Overall, looks great! Awesome feature!

A couple of bugs/thoughts:
1) On vertical funnels, if you rename later steps, the custom names don't appear in the graph below:
![image](https://user-images.githubusercontent.com/4813045/134738506-0c689a1f-2083-483b-b7c7-17fa50d3a8c0.png)

2) If I rename the first step, and then I add additional steps, the new steps automatically get the custom name of the first step. I'd expect the new steps to have the default name (e.g. $pageview)

3) It doesn't seem like we use the word "query step" elsewhere in the product. I wonder if we should keep it consistent with "graph series" and "funnel step" on the respective page.
![image](https://user-images.githubusercontent.com/4813045/134738982-dd647f7e-d8da-4153-83b8-69895a18a5f0.png)

4) On funnels, it might make sense to include the custom names in the "time to convert" and "Historical trends" drop downs.
![image](https://user-images.githubusercontent.com/4813045/134739697-3dd7c0a3-9fe0-424a-a3bb-c30136d2183b.png)

5) I had some funny behavior around custom names not appearing after saving an insight, but I can't repro' it so.... maybe just a fluke.

rcmarron commented on frontend/src/lib/components/EntityFilterInfo.tsx:
```diff
@@ -0,0 +1,41 @@
+        return <TextWrapper title="Select filter">Select filter</TextWrapper>
+    }
+
+    const titleTranslation = getKeyMapping(title, 'event')?.label ?? title ?? undefined
+    const subTitleTranslation = getKeyMapping(subtitle, 'event')?.label ?? subtitle ?? undefined
```
Translation feels like a misnomer here (particularly with strings). Maybe `titleToDisplay`?

alexkim205 replied:
sounds much better. english is hard

alexkim205 commented:
Thanks for the comprehensive review Rick!

> 1. On vertical funnels, if you rename later steps, the custom names don't appear in the graph below:

> 5. I had some funny behavior around custom names not appearing after saving an insight, but I can't repro' it so.... maybe just a fluke.

I think this might be happening bc these filter names are cached on the backend (which could also be related to point 5). Could you try hard refreshing the graph a bunch of times and see if this fixes things?

> 2. If I rename the first step, and then I add additional steps, the new steps automatically get the custom name of the first step. I'd expect the new steps to have the default name (e.g. $pageview)

Good catch, fixed!

> 3. It doesn't seem like we use the word "query step" elsewhere in the product. I wonder if we should keep it consistent with "graph series" and "funnel step" on the respective page.

Changed!

> 4. On funnels, it might make sense to include the custom names in the "time to convert" and "Historical trends" drop downs.

Good call. I think it'd be good to include the custom names in the dropdown menu. I'll exclude them in the input field itself to avoid making that label variably longer.

alexkim205 committed:
fix a few bugs

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/entity-filter-renaming

alexkim205 commented:
As this already went through one round of code reviews, I'm going to merge this one in for now so that it's not out for too long. I'll address any followup bugs in future PR's so feel free to keep the convo going here.

joesaunderson commented:
Love this!!

mariusandra commented on frontend/src/scenes/insights/ActionFilter/ActionFilter.tsx:
```diff
@@ -94,6 +97,8 @@ export const ActionFilter = React.forwardRef<HTMLDivElement, ActionFilterProps>(
         const { localFilters } = useValues(logic)
         const { addFilter, setLocalFilters } = useActions(logic)
 
+        const [modalOpen, setModalOpen] = useState(false)
+
         // No way around this. Somehow the ordering of the logic calling each other causes stale "localFilters"
```
Can't this state be moved into the entity filter logic?

mariusandra commented on frontend/src/scenes/insights/ActionFilter/FilterRenameModal.tsx:
```diff
@@ -0,0 +1,81 @@
+    const { selectedFilter } = useValues(entityFilterLogic)
+    const { renameFilter, selectFilter } = useActions(entityFilterLogic)
+
+    const [name, setName] = useState(getDisplayNameFromEntityFilter(selectedFilter) ?? '')
+
+    useEffect(() => {
+        setName(getDisplayNameFromEntityFilter(selectedFilter) ?? '')
+    }, [selectedFilter])
```
I'd really love to see this refactored into a logic... plus tests, etc.

mariusandra reviewed (commented):
Retroactive review. All seems fine, though I'd really love to see the modal shown/hidden info to move into a logic as well. `entityFilterLogic` perhaps. I keep repeating myself, but if even for little local states like this, we store them in a logic, we get great benefits. Such as not having to rewrite this later when asked, having all the logic in an app follow the same pattern, easier testing, no prop passing, etc...

alexkim205 commented:
Yup I can move this into a logic and add some tests as well. Tracking here #6153
